File: 2024/10 October/Split Linked List Alternatingly/main.py
# Link: https://www.geeksforgeeks.org/problems/split-singly-linked-list-alternatingly/1


class Node:

    def __init__(self, x):
        self.data = x
        self.next = None


def printList(node):
    while node is not None:
        print(node.data, end=" ")
        node = node.next
    print()


def createLinkedList(arr):
    head = Node(arr[0])
    temp = head
    for i in range(1, len(arr)):
        temp.next = Node(arr[i])
        temp = temp.next
    return head


class Solution:
    def alternatingSplitList(self, head):
        if head is None:
            return [None, None]
        odd_dummy = Node(0)
        even_dummy = Node(0)
        odd = odd_dummy
        even = even_dummy
        curr = head
        o = True
        while curr is not None:
            if o:
                odd.next = curr
                odd = odd.next
            else:
                even.next = curr
                even = even.next
            curr = curr.next
            o = not o
        odd.next = None
        even.next = None
        return [odd_dummy.next, even_dummy.next]


if __name__ == "__main__":
    arr = [0, 1, 0, 1, 0, 1]
    head = createLinkedList(arr)
    ob = Solution()
    ans = ob.alternatingSplitList(head)
    printList(ans[0])
    printList(ans[1])

    arr = [2, 5, 8, 9, 6]
    head = createLinkedList(arr)
    ans = ob.alternatingSplitList(head)
    printList(ans[0])
    printList(ans[1])

File: 2024/10 October/Split Linked List Alternatingly/main.cpp
// Link: https://www.geeksforgeeks.org/problems/split-singly-linked-list-alternatingly/1

#include <iostream>
#include <vector>
using namespace std;

struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

void printList(struct Node *node)
{
    while (node != NULL)
    {
        cout << node->data << " ";
        node = node->next;
    }
    cout << endl;
}

struct Node *createLinkedList(vector<int> &arr)
{
    struct Node *head = new Node(arr[0]);
    struct Node *temp = head;
    for (int i = 1; i < arr.size(); i++)
    {
        temp->next = new Node(arr[i]);
        temp = temp->next;
    }
    return head;
}

class Solution
{
public:
    vector<Node *> alternatingSplitList(struct Node *head)
    {
        if (head == NULL)
            return {NULL, NULL};
        Node *odddummy = new Node(0);
        Node *evendummy = new Node(0);
        Node *odd = odddummy;
        Node *even = evendummy;
        Node *curr = head;
        bool o = true;
        while (curr != NULL)
        {
            if (o)
            {
                odd->next = curr;
                odd = odd->next;
            }
            else
            {
                even->next = curr;
                even = even->next;
            }
            curr = curr->next;
            o = !o;
        }
        odd->next = NULL;
        even->next = NULL;
        return {odddummy->next, evendummy->next};
    }
};

int main()
{
    vector<int> arr = {0, 1, 0, 1, 0, 1};
    struct Node *head = createLinkedList(arr);
    Solution ob;
    vector<Node *> ans = ob.alternatingSplitList(head);
    printList(ans[0]);
    printList(ans[1]);

    arr = {2, 5, 8, 9, 6};
    head = createLinkedList(arr);
    ans = ob.alternatingSplitList(head);
    printList(ans[0]);
    printList(ans[1]);
    return 0;
}